### ISSUE TITLE
WebSockets support

### ISSUE THREAD

**Issue by mzimmerman** [elazarl/goproxy#72]
From a brief look, goproxy does not support proxying websockets (in transparent mode at least)

http://www.websocket.org/echo.html
## From chromium

WebSocket connection to 'ws://echo.websocket.org/?encoding=text' failed: Error during WebSocket handshake: Unexpected response code: 400
## From goproxy

2014/11/07 09:41:19 [120] INFO: Got request / echo.websocket.org GET http://echo.websocket.org/?encoding=text
2014/11/07 09:41:19 [120] INFO: Sending request GET http://echo.websocket.org/?encoding=text
2014/11/07 09:41:19 [120] INFO: Received response 400 WebSocket Upgrade Failure
2014/11/07 09:41:19 [120] INFO: Copying response to client 400 WebSocket Upgrade Failure [400]
2014/11/07 09:41:19 [120] INFO: Copied 77 bytes to client error=<nil>

**Comment by elazarl** [comment]
@mzimmerman summarizing the issue, in order to support websockets in transparent proxy, we need to:

1. Recognize upgrade requests.
2. In case of an upgrade request, hijack the connection, and stream all TCP data to the remote host.

I'm not sure whether we need to include that in the core library. I think we simply allow user to hijack connection, and he'll do whatever he wishes in a filter.

How many people are using goproxy in transparent setting?

**Comment by k2xl** [comment]
has there been any update to this issue? lots of the web is turning to websockets, so it would be great if goproxy could do that interception

**Comment by elazarl** [comment]
@k2xl are you using transparent proxy?

**Comment by k2xl** [comment]
No, I'm modifying the requests (I'm assuming that is what you mean by transparent proxy)

**Comment by elazarl** [comment]
@k2xl I meant, do tell your browser to send explicit proxy request. Do you set in the browser's settings goproxy as the HTTP+HTTPS proxy?

**Comment by k2xl** [comment]
Yes

**Comment by 0xbkt** [comment]
Hello? WebSockets are not working when `goproxy.AlwaysMitm`'ing requests, at least in my case.
